# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [48]:
# TODO: import the pandas library
import pandas as pd
# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=4, dtype = 'str')
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=4, dtype = 'str')

# TODO: remove the 'Unnamed:62' column from each data set
df_rural.drop(columns = 'Unnamed: 62', inplace = True)
df_electricity.drop(columns = 'Unnamed: 62', inplace = True)
# TODO: combine the two data sets together using the concat method

In [49]:
pd.concat([df_rural,df_electricity])

Country Name Country Code                            Indicator Name  \
0           Aruba          ABW  Rural population (% of total population)   
1     Afghanistan          AFG  Rural population (% of total population)   
2          Angola          AGO  Rural population (% of total population)   
3         Albania          ALB  Rural population (% of total population)   
4         Andorra          AND  Rural population (% of total population)   
..            ...          ...                                       ...   
259        Kosovo          XKX   Access to electricity (% of population)   
260   Yemen, Rep.          YEM   Access to electricity (% of population)   
261  South Africa          ZAF   Access to electricity (% of population)   
262        Zambia          ZMB   Access to electricity (% of population)   
263      Zimbabwe          ZWE   Access to electricity (% of population)   

     Indicator Code    1960    1961    1962    1963    1964    1965  ...  \
0    SP.RUR.TOTL.ZS  49.224  49.239  49.254   49.27  49.285    49.3  ...   
1    SP.RUR.TOTL.ZS  91.779  91.492  91.195   90.89  90.574   90.25  ...   
2    SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...   
3    SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842   68.77  ...   
4    SP.RUR.TOTL.ZS   41.55  39.017  36.538  34.128  31.795  29.555  ...   
..              ...     ...     ...     ...     ...     ...     ...  ...   
259  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
260  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
261  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
262  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   
263  EG.ELC.ACCS.ZS     NaN     NaN     NaN     NaN     NaN     NaN  ...   

                 2008              2009              2010              2011  \
0              56.217            56.579            56.941            57.302   
1              76.054            75.687            75.311            74.926   
2              61.491            60.701            59.903              59.1   
3              50.009            48.924            47.837            46.753   
4              11.133            11.648            12.183             12.74   
..                ...               ...               ...               ...   
259  97.7946701049805  98.3770599365234                 0  99.4304351806641   
260  60.3865394592285  61.7513999938965  63.1342811584473  64.5321807861328   
261              81.9              82.7              82.9              84.7   
262  22.7146854400635  23.2356491088867                22  24.3286437988281   
263  35.4573097229004  43.3690818728747  36.0630569458008              36.9   

                 2012              2013              2014              2015  \
0              57.636            57.942            58.221            58.472   
1              74.532            74.129            73.718            73.297   
2              58.301             57.51            56.726             55.95   
3               45.67            44.617            43.591            42.593   
4              13.292            13.835            14.367            14.885   
..                ...               ...               ...               ...   
259  99.7736511230469  99.9425048828125  99.9931335449219               100   
260  65.9421005249023              75.6              66.1  70.2139053344727   
261              85.3              85.4                86              85.5   
262  24.8946666717529  25.4697017669678              27.9              31.1   
263  36.7288780212402  37.0768127441406              32.3              33.7   

                 2016    2017  
0              58.696  58.893  
1              72.868   72.43  
2              55.181  54.422  
3              41.624  40.684  
4              15.388  15.873  
..                ...     ...  
259               100     NaN  
260  71.6423492431641    

In [50]:
df_electricity

Country Name Country Code                           Indicator Name  \
0           Aruba          ABW  Access to electricity (% of population)   
1     Afghanistan          AFG  Access to electricity (% of population)   
2          Angola          AGO  Access to electricity (% of population)   
3         Albania          ALB  Access to electricity (% of population)   
4         Andorra          AND  Access to electricity (% of population)   
..            ...          ...                                      ...   
259        Kosovo          XKX  Access to electricity (% of population)   
260   Yemen, Rep.          YEM  Access to electricity (% of population)   
261  South Africa          ZAF  Access to electricity (% of population)   
262        Zambia          ZMB  Access to electricity (% of population)   
263      Zimbabwe          ZWE  Access to electricity (% of population)   

     Indicator Code 1960 1961 1962 1963 1964 1965  ...              2008  \
0    EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...   93.086166381836   
1    EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...              42.4   
2    EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...  31.2680130004883   
3    EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...               100   
4    EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...               100   
..              ...  ...  ...  ...  ...  ...  ...  ...               ...   
259  EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...  97.7946701049805   
260  EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...  60.3865394592285   
261  EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...              81.9   
262  EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...  22.7146854400635   
263  EG.ELC.ACCS.ZS  NaN  NaN  NaN  NaN  NaN  NaN  ...  35.4573097229004   

                 2009              2010              2011              2012  \
0    93.3545455932617  93.3562921509542  93.9423751831055  94.2558135986328   
1    44.8548851013184              42.7  43.2220189082037              69.1   
2    32.3824691772461   33.514949798584              34.6   35.821964263916   
3                 100               100               100               100   
4                 100               100               100               100   
..                ...               ...               ...               ...   
259  98.3770599365234                 0  99.4304351806641  99.7736511230469   
260  61.7513999938965  63.1342811584473  64.5321807861328  65.9421005249023   
261              82.7              82.9              84.7              85.3   
262  23.2356491088867                22  24.3286437988281  24.8946666717529   
263  43.3690818728747  36.0630569458008              36.9  36.7288780212402   

                 2013              2014              2015              2016  \
0    94.5782623291016  94.9067230224609  95.2381820678711  95.5701446533203   
1    67.2595520019531              89.5              71.5  84.1371383666992   
2    36.9904899597168                32                42  40.5206069946289   
3                 100               100               100               100   
4                 100               100               100               100   
..                ...               ...               ...               ...   
259  99.9425048828125  99.9931335449219               100               100   
260              75.6              66.1  70.2139053344727  71.6423492431641   
261              85.4                86              85.5              84.2   
262  25.4697017669678              27.9              31.1  27.2193374633789   
263  37.0768127441406              32.3              33.7  38.1451377868652   

    2017  
0    NaN  
1    NaN  
2    NaN  
3    NaN  
4    NaN  
..   ...  
259  NaN  
260  NaN  
261  NaN  
262  NaN  
263  NaN  

[264 rows x 62 columns]

In [51]:
with open('electricity_access_percent.csv') as f:
    for i in range(5):
        read_data = f.readline()
        print(read_data)

"Data Source","World Development Indicators",



"Last Updated Date","2018-06-28",



"Country Name","Country Code","Indicator Name","Indicator Code","1960","1961","1962","1963","1964","1965","1966","1967","1968","1969","1970","1971","1972","1973","1974","1975","1976","1977","1978","1979","1980","1981","1982","1983","1984","1985","1986","1987","1988","1989","1990","1991","1992","1993","1994","1995","1996","1997","1998","1999","2000","2001","2002","2003","2004","2005","2006","2007","2008","2009","2010","2011","2012","2013","2014","2015","2016","2017",



# Exercise 2 (Challenge)

This exercise is more challenging.

The resulting data frame should look like this:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [63]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value
df_rural_mod = df_rural.drop(columns = ['Indicator Name', 'Indicator Code'])
df_rural_mod = df_rural_mod.melt(id_vars=['Country Name','Country Code'], var_name = 'Year', value_name = 'Rural_Value')
df_electricity_mod = df_electricity.drop(columns = ['Indicator Name', 'Indicator Code'])
df_electricity_mod = df_electricity_mod.melt(id_vars=['Country Name','Country Code'], var_name = 'Year', value_name = 'Electricity_Value')


# TODO: drop any columns from the data frames that aren't needed

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year

# TODO: sort the results by country and then by year

df_combined = pd.merge(df_rural_mod,df_electricity_mod, on=['Country Name','Country Code','Year'])

df_combined.sort_values(by=['Country Name', 'Year'])

Country Name Country Code  Year Rural_Value Electricity_Value
1      Afghanistan          AFG  1960      91.779               NaN
265    Afghanistan          AFG  1961      91.492               NaN
529    Afghanistan          AFG  1962      91.195               NaN
793    Afghanistan          AFG  1963       90.89               NaN
1057   Afghanistan          AFG  1964      90.574               NaN
...            ...          ...   ...         ...               ...
14255     Zimbabwe          ZWE  2013      67.346  37.0768127441406
14519     Zimbabwe          ZWE  2014      67.499              32.3
14783     Zimbabwe          ZWE  2015      67.624              33.7
15047     Zimbabwe          ZWE  2016      67.723  38.1451377868652
15311     Zimbabwe          ZWE  2017      67.793               NaN

[15312 rows x 5 columns]

In [53]:
df_electricity_mod.dtypes

Country Name         object
Country Code         object
Year                 object
Electricity_Value    object
dtype: object

In [54]:
df_electricity_mod.dtypes

Country Name         object
Country Code         object
Year                 object
Electricity_Value    object
dtype: object

In [58]:
df_total = 

In [60]:
df_total[df_total['Country Name']== 'Aruba']

Country Name Country Code  Year Rural_Value Electricity_Value
0            Aruba          ABW  1960      49.224               NaN
264          Aruba          ABW  1961      49.239               NaN
528          Aruba          ABW  1962      49.254               NaN
792          Aruba          ABW  1963       49.27               NaN
1056         Aruba          ABW  1964      49.285               NaN
1320         Aruba          ABW  1965        49.3               NaN
1584         Aruba          ABW  1966      49.315               NaN
1848         Aruba          ABW  1967       49.33               NaN
2112         Aruba          ABW  1968      49.346               NaN
2376         Aruba          ABW  1969      49.361               NaN
2640         Aruba          ABW  1970      49.376               NaN
2904         Aruba          ABW  1971      49.391               NaN
3168         Aruba          ABW  1972      49.407               NaN
3432         Aruba          ABW  1973      49.422               NaN
3696         Aruba          ABW  1974      49.437               NaN
3960         Aruba          ABW  1975      49.452               NaN
4224         Aruba          ABW  1976      49.468               NaN
4488         Aruba          ABW  1977      49.483               NaN
4752         Aruba          ABW  1978      49.498               NaN
5016         Aruba          ABW  1979      49.513               NaN
5280         Aruba          ABW  1980      49.528               NaN
5544         Aruba          ABW  1981      49.544               NaN
5808         Aruba          ABW  1982      49.559               NaN
6072         Aruba          ABW  1983      49.574               NaN
6336         Aruba          ABW  1984      49.589               NaN
6600         Aruba          ABW  1985      49.605               NaN
6864         Aruba          ABW  1986       49.62               NaN
7128         Aruba          ABW  1987      49.635               NaN
7392         Aruba          ABW  1988       49.65               NaN
7656         Aruba          ABW  1989      49.665               NaN
7920         Aruba          ABW  1990      49.681  88.4453506469727
8184         Aruba          ABW  1991      49.696  88.7808456420898
8448         Aruba          ABW  1992      50.002  89.1158294677734
8712         Aruba          ABW  1993      50.412    89.44775390625
8976         Aruba          ABW  1994      50.823  89.7735595703125
9240         Aruba          ABW  1995      51.233  90.0901870727539
9504         Aruba          ABW  1996      51.644  90.3945846557617
9768         Aruba          ABW  1997      52.054  90.6836776733398
10032        Aruba          ABW  1998      52.464  90.9544219970703
10296        Aruba          ABW  1999      52.873  91.2037506103516
10560        Aruba          ABW  2000      53.283  91.6603980031457
10824        Aruba          ABW  2001      53.661  91.6380920410156
11088        Aruba          ABW  2002      54.028  91.8337173461914
11352        Aruba          ABW  2003      54.394  92.0230484008789
11616        Aruba          ABW  2004       54.76  92.2121658325195
11880        Aruba          ABW  2005      55.125  92.4071197509765
12144        Aruba          ABW  2006      55.489  92.6139831542969
12408        Aruba          ABW  2007      55.853  92.8388214111328
12672        Aruba          ABW  2008      56.217   93.086166381836
12936        Aruba          ABW  2009      56.579  93.3545455932617
13200        Aruba          ABW  2010      56.941  93.3562921509542
13464        Aruba          ABW  2011      57.302  93.9423751831055
13728        Aruba          ABW  2012      57.636  94.2558135986328
13992        Aruba          ABW  2013      57.942  94.5782623291016
14256        Aruba          ABW  2014      58.221  94.9067230224609
14520        Aruba          ABW  2015      58.472  95.2381820678711
14784        Aruba          ABW  2016      58.696  95.5701446533203
15048        Aruba          ABW  2017      58.893             